File: data/src/main/java/com/joxad/zikobot/data/module/localmusic/manager/LocalMusicManager.java
package com.joxad.zikobot.data.module.localmusic.manager;

import android.content.ContentResolver;
import android.content.ContentUris;
import android.content.Context;
import android.database.Cursor;
import android.database.CursorIndexOutOfBoundsException;
import android.net.Uri;
import android.provider.MediaStore;

import com.joxad.zikobot.data.db.model.ZikoAlbum;
import com.joxad.zikobot.data.db.model.ZikoAlbum_Table;
import com.joxad.zikobot.data.db.model.ZikoArtist;
import com.joxad.zikobot.data.db.model.ZikoArtist_Table;
import com.joxad.zikobot.data.db.model.ZikoTrack;
import com.joxad.zikobot.data.db.model.ZikoTrack_Table;
import com.joxad.zikobot.data.module.localmusic.model.LocalTrack;
import com.raizlabs.android.dbflow.sql.language.Select;

import java.io.File;

import io.reactivex.Observable;
import io.reactivex.subjects.BehaviorSubject;


/**
 * Created by Josh on 06/04/2016.
 */

public enum LocalMusicManager {
    INSTANCE;

    private Context context;

    public BehaviorSubject<Boolean> synchroDone;
    private ContentResolver cr;

    public void init(Context context) {
        this.context = context;
        synchroDone = BehaviorSubject.create();
        cr = context.getContentResolver();

    }

    public Observable<Boolean> observeSynchro() {
        return Observable.fromCallable(() -> {
            syncLocalData();
            synchroDone.onNext(true);
            return true;
        });
    }


    public void syncLocalData() {
        for (LocalTrack localTrack : TrackLoader.getAllTracks(context)) {
            ZikoArtist zikoArtist = createArtistIfNeed(localTrack.getArtistId(), localTrack.getArtistName());
            ZikoAlbum zikoAlbum = createAlbumIfNeed(localTrack.getAlbumId(), localTrack.getAlbumName(), zikoArtist);
            createTrack(localTrack, zikoArtist, zikoAlbum);
        }
    }


    private ZikoArtist createArtistIfNeed(long artistId, String artistName) {
        ZikoArtist zikoArtist = new Select().from(ZikoArtist.class).where(ZikoArtist_Table.localId.eq(artistId)).querySingle();
        if (zikoArtist == null) {
            zikoArtist = ZikoArtist.Companion.local(artistId, artistName);
            zikoArtist.save();
        }
        return zikoArtist;
    }

    final public static Uri sArtworkUri = Uri
            .parse("content://media/external/audio/albumart");

    private ZikoAlbum createAlbumIfNeed(long albumId, String albumName, ZikoArtist zikoArtist) {
        ZikoAlbum zikoAlbum = new Select().from(ZikoAlbum.class)
                .where(ZikoAlbum_Table.localId.eq(albumId))
                .or(ZikoAlbum_Table.name.like(albumName))
                .querySingle();
        if (zikoAlbum == null) {
            zikoAlbum = ZikoAlbum.Companion.local(albumId, albumName, zikoArtist);
            Uri albumArtUri = ContentUris.withAppendedId(sArtworkUri, albumId);
            if (isUriBitmapValid(albumArtUri))
                zikoAlbum.image = albumArtUri.toString();
            zikoAlbum.save();
        }
        return zikoAlbum;
    }


    private boolean isUriBitmapValid(Uri uri) {
        String[] projection = {MediaStore.MediaColumns.DATA};
        Cursor cur = cr.query(uri, projection, null, null, null);
        boolean b = false;
        if (cur != null) {
            cur.moveToFirst();
            try {

                String filePath = cur.getString(0);
                if (filePath == null || filePath.isEmpty()) {
                    b = false;
                } else b = (new File(filePath)).exists();
            } catch (CursorIndexOutOfBoundsException e) {
                b = false;
            }


        } else {
            b = false;
        }
        return b;
    }

    private ZikoTrack createTrack(LocalTrack localTrack, ZikoArtist artist, ZikoAlbum album) {
        ZikoTrack zikoTrack = new Select().from(ZikoTrack.class).where(ZikoTrack_Table.localId.eq(localTrack.getId())).querySingle();
        if (zikoTrack == null) {
            zikoTrack = ZikoTrack.local(localTrack, artist, album);
            zikoTrack.save();
        }
        return zikoTrack;
    }
}

File: data/src/main/java/com/joxad/zikobot/data/module/localmusic/manager/LocalMusicManager.kt
package com.joxad.zikobot.data.module.localmusic.manager

import android.content.ContentResolver
import android.content.ContentUris
import android.content.Context
import android.database.Cursor
import android.database.CursorIndexOutOfBoundsException
import android.net.Uri
import android.provider.MediaStore
import android.util.Log

import com.joxad.zikobot.data.db.model.ZikoAlbum
import com.joxad.zikobot.data.db.model.ZikoAlbum_Table
import com.joxad.zikobot.data.db.model.ZikoArtist
import com.joxad.zikobot.data.db.model.ZikoArtist_Table
import com.joxad.zikobot.data.db.model.ZikoTrack
import com.joxad.zikobot.data.db.model.ZikoTrack_Table
import com.joxad.zikobot.data.module.localmusic.model.LocalTrack
import com.joxad.zikobot.data.module.spotify_api.manager.SpotifyApiManager
import com.joxad.zikobot.data.module.spotify_api.model.SpotifyResultTracks
import com.joxad.zikobot.data.module.spotify_api.model.SpotifyTrack
import com.raizlabs.android.dbflow.sql.language.Select

import java.io.File

import io.reactivex.Observable
import io.reactivex.ObservableSource
import io.reactivex.Single
import io.reactivex.functions.Function
import io.reactivex.subjects.PublishSubject


/**
 * Created by Josh on 06/04/2016.
 */

class LocalMusicManager(private val context: Context) {
    private val synchroDone: PublishSubject<Boolean>
    private val cr: ContentResolver

    init {
        synchroDone = PublishSubject.create()
        cr = context.contentResolver
    }

    fun observeSynchro(): Observable<Boolean> {
        return Observable.fromCallable {
            try {
                syncLocalData()
            } catch (e: Exception) {
                Log.d(LocalMusicManager::class.java.simpleName, e.localizedMessage)
            }

            synchroDone.onNext(true)
            true
        }
    }


    fun syncLocalData() {
        try {
            for (localTrack in TrackLoader.getAllTracks(context)) {
                val zikoArtist = createArtistIfNeed(localTrack.artistId, localTrack.artistName)
                val zikoAlbum = createAlbumIfNeed(localTrack.albumId.toLong(), localTrack.albumName, zikoArtist)
                createTrack(localTrack, zikoArtist, zikoAlbum)
            }
        } catch (e: Exception) {
            Log.e(LocalMusicManager::class.java.simpleName, e.localizedMessage)
        }

    }

    private fun createArtistIfNeed(artistId: Long, artistName: String?): ZikoArtist {
        var zikoArtist = Select().from(ZikoArtist::class.java).where(ZikoArtist_Table.localId.eq(artistId)).querySingle()
        if (zikoArtist == null) {
            zikoArtist = ZikoArtist.local(artistId, artistName!!)
            zikoArtist.favorite = true
            zikoArtist.save()
        }
        return zikoArtist
    }

    private fun createAlbumIfNeed(albumId: Long, albumName: String?, zikoArtist: ZikoArtist): ZikoAlbum {
        var zikoAlbum = Select().from(ZikoAlbum::class.java)
                .where(ZikoAlbum_Table.localId.eq(albumId))
                .or(ZikoAlbum_Table.name.like(albumName!!))
                .querySingle()
        if (zikoAlbum == null) {
            zikoAlbum = ZikoAlbum.local(albumId, albumName, zikoArtist)
            val albumArtUri = ContentUris.withAppendedId(sArtworkUri, albumId)
            if (isUriBitmapValid(albumArtUri))
                zikoAlbum.image = albumArtUri.toString()
            zikoAlbum.favorite = true
            zikoAlbum.save()
        }
        return zikoAlbum
    }


    private fun isUriBitmapValid(uri: Uri): Boolean {
        val projection = arrayOf(MediaStore.MediaColumns.DATA)
        val cur = cr.query(uri, projection, null, null, null)
        var b = false
        if (cur != null) {
            cur.moveToFirst()
            try {

                val filePath = cur.getString(0)
                if (filePath == null || filePath.isEmpty()) {
                    b = false
                } else
                    b = File(filePath).exists()
            } catch (e: CursorIndexOutOfBoundsException) {
                b = false
            }


        } else {
            b = false
        }
        return b
    }

    private fun createTrack(localTrack: LocalTrack, artist: ZikoArtist, album: ZikoAlbum): ZikoTrack {
        var zikoTrack = Select().from(ZikoTrack::class.java).where(ZikoTrack_Table.localId.eq(localTrack.id)).querySingle()
        if (zikoTrack == null) {
            zikoTrack = ZikoTrack.local(localTrack, artist, album)
            zikoTrack!!.save()
        }
        return zikoTrack
    }

    fun synchroDoneObservable(): Single<Boolean> {
        return synchroDone.lastOrError()
    }

    companion object {

        val sArtworkUri = Uri
                .parse("content://media/external/audio/albumart")
    }
}
